### PR TITLE
WT-7348 CMake Complete POSIX support fixes

### PR DESCRIPTION
**Pull request by alisonfel**
This PR appends some additional fixes to the final CMake POSIX support, to be merged with all the other changes under WT-7348.
More specifically, this PR adds:
* Fixes to `define_csuite_test` helper so we can run csuite tests in parallel without them running over each other
* Adjustments to the following C tests:
   *  Prevent `wt2909_checkpoint_integrity` & `wt3120_filesys` from assuming the build location of the fail fs extension
library/module
   * Avoid `syscall` test assuming the layout of `build_posix`
* Avoid compiling with `fPIC` for static builds
*  Respect CMake Build Modes: Making it IDE friendly
*  ASAN and UBSan build mode

### CODE REVIEW

File: build_cmake/configs/modes.cmake [hunk 1]
```diff
@@ -8,8 +8,56 @@
 
 # Establishes build configuration modes we can use when compiling.
 
+# Create an ASAN build variant
```
Comment (alisonfel) [inline build_cmake/configs/modes.cmake:11]: These changes were a slight scope creep and can take it out of the PR if you rather it not be added.
Thought it might be useful to add some additional build modes to make it super easy to run the build+tests under ASan and UBSan (I personally always forget the flags).

Comment (tetsuo-cpp) [reply build_cmake/configs/modes.cmake:11]: I'd definitely like to keep these. That's handy.

File: test/utility/misc.c [hunk 2]
```diff
@@ -141,11 +141,19 @@ testutil_clean_work_dir(const char *dir)
  *     Get the git top level directory and concatenate the build directory.
  */
 void
-testutil_build_dir(char *buf, int size)
+testutil_build_dir(TEST_OPTS *opts, char *buf, int size)
 {
     FILE *fp;
     char *p;
 
+    /* If a build directory was manually given as an option we can directly
```
Comment (tetsuo-cpp) [inline test/utility/misc.c:149]: Nit: Comment should be like:
```
/*
 * Blah blah blah blah blah
 * blah blah
 */
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -86,6 +90,15 @@ target_compile_options(
     PRIVATE ${COMPILER_DIAGNOSTIC_FLAGS}
 )
 
+if(ENABLE_STATIC)
+    # Avoid compiling with fPIC for static builds, this can often lead to
+    # GCC disabling many optimizations e.g. inlining, possibly introducing
+    # an overhead.
+    set_property(TARGET wiredtiger PROPERTY POSITION_INDEPENDENT_CODE OFF)
+else()
+    set_property(TARGET wiredtiger PROPERTY POSITION_INDEPENDENT_CODE ON)
```
Comment (tetsuo-cpp) [inline CMakeLists.txt:99]: What happens if you build a shared lib with `-fPIC` anyway? Do bad things happen.

File: test/csuite/wt4333_handle_locks/main.c [hunk 4]
```diff
@@ -320,23 +322,30 @@ run(int argc, char *argv[])
...
         default:
             fprintf(stderr, "usage: %s [-v]\n", argv[0]);
             return (EXIT_FAILURE);
         }
     }
-
     (void)signal(SIGALRM, on_alarm);
 
+    if (default_home)
+        testutil_work_dir_from_path(home, sizeof(home), "WT_TEST.wt4333_handle_locks");
```
Comment (tetsuo-cpp) [inline test/csuite/wt4333_handle_locks/main.c:347]: How come this one is needed? The others changes make sense to me because I think `.libs` is a LibTool thing, but not sure about this one.

**Review by tetsuo-cpp — APPROVED** [review]
LGTM modulo my comment formatting nitpick.

File: test/utility/misc.c (hunk 2 above)
Comment (alisonfel) [reply test/utility/misc.c]: Thanks for pointing that out! I ran through the dist tooling anyway and it converted it to a single-line comment. Pushed the formatting update :)

File: CMakeLists.txt (hunk 3 above)
Comment (alisonfel) [reply CMakeLists.txt:99]: We will unfortunately generate dreaded linker errors, failing to compile the shared library. They are typically of the type `relocation R_X86_64_PC32 against symbol ...`. This is mainly associated with our large number of`extern` functions found in `src/include/extern.h`. We need to compile with PIC so we can call external symbols (function calls across shared-object boundaries).

Currently we depend on libtool to do all the PIC magic, which is why we don't currently see with `-fPIC` in our compilation output (libtool generates object files `libs/*.lo`file for each object. This managing PIC objects when performing the final linker step of shared library). 
Since we are not using libtool, we need to ensure `-fPIC` is present when compiling a shared library with CMake.

File: test/csuite/wt4333_handle_locks/main.c (hunk 4 above)
Comment (alisonfel) [reply test/csuite/wt4333_handle_locks/main.c:347]: Oh right, I added those specific lines for backwards compatibility with autoconf, since `make check` won't call the test with a `-h` flag. Essentially allowing it to invoke the test as it currently does. Most of the csuite tests will accept the `-h` flags as they delegate all the option parsing to the testutil library `test/utility/parse_opts.c`. Though this specific test doesn't, introducing its own options.
(for context, I added the `-h` so CMake can declare a working directory for its corresponding CTest command, given it filters all the test output to a specific binary directory.)

Comment (tetsuo-cpp) [reply test/csuite/wt4333_handle_locks/main.c:347]: Awesome, that makes sense.

**Review by lukech — APPROVED** [review]

**Review by kommiharibabu — APPROVED** [review]
LGTM
